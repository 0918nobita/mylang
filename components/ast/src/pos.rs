use std::fmt;

use serde::{Deserialize, Serialize};

#[derive(Clone, Default, Serialize, Deserialize)]
pub struct Pos {
    line: u32,
    column: u32,
}

impl fmt::Debug for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line, self.column)
    }
}
