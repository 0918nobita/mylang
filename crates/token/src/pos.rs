use std::fmt;

use serde::{Deserialize, Serialize};

#[derive(Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Pos {
    line: u32,
    character: u32,
}

impl Pos {
    pub fn new(line: u32, character: u32) -> Self {
        Self { line, character }
    }

    /// 次の文字位置に移動する (改行ではない文字が続くことを想定している)
    ///
    /// ```
    /// use token::Pos;
    ///
    /// let mut pos = Pos::new(3, 7);
    /// pos.next_char();
    ///
    /// let expected = Pos::new(3, 8);
    ///
    /// assert_eq!(pos, expected);
    /// ```
    pub fn next_char(&mut self) {
        self.character += 1;
    }

    /// 次の行の先頭に移動する
    ///
    /// ```
    /// use token::Pos;
    ///
    /// let mut pos = Pos::new(1, 8);
    /// pos.next_line();
    ///
    /// let expected = Pos::new(2, 0);
    /// assert_eq!(pos, expected);
    /// ```
    pub fn next_line(&mut self) {
        self.line += 1;
        self.character = 0;
    }
}

impl fmt::Debug for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line, self.character)
    }
}

impl fmt::Display for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line + 1, self.character + 1)
    }
}
