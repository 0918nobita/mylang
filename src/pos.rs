use std::fmt;

#[derive(Default)]
pub struct Pos {
    line: u32,
    column: u32,
}

impl fmt::Debug for Pos {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.line, self.column)
    }
}
