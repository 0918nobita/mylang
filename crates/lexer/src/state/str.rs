use mylang_token::{range, Pos, Token};

use crate::{LexErr, LexResult};

#[derive(Clone, Debug, Default)]
pub struct StrState {
    start: Pos,
    pub end: Pos,
    pub escape: bool,
    acc: String,
}

impl StrState {
    pub fn new(pos: Pos) -> Self {
        Self {
            start: pos.clone(),
            end: pos,
            escape: false,
            acc: String::new(),
        }
    }

    pub fn try_append_char(&self, pos: Pos, c: char) -> LexResult<Self> {
        match (self.escape, c) {
            (_, '\n') => Err(LexErr::ForbiddenChar(pos, c)),

            (true, c @ ('n' | '\\' | '"')) => Ok(Self {
                start: self.start.clone(),
                end: pos,
                escape: false,
                acc: format!("{}{c}", self.acc),
            }),

            (true, _) => Err(LexErr::InvalidEscapeSequence(pos, c)),

            (false, '\\') => Ok(Self {
                start: self.start.clone(),
                end: pos,
                escape: true,
                acc: self.acc.clone(),
            }),

            (false, c) => Ok(Self {
                start: self.start.clone(),
                end: pos,
                escape: false,
                acc: format!("{}{c}", self.acc),
            }),
        }
    }

    pub fn tokenize(&self) -> Token {
        if self.escape {
            panic!("Illigal state")
        }
        Token::Str(
            range!(self.start.clone(), self.end.clone()),
            self.acc.clone(),
        )
    }
}

#[cfg(test)]
mod tests {
    use mylang_token::pos;

    use super::StrState;

    #[test]
    fn newline() {
        let state = StrState::default();
        let state = state.try_append_char(pos!(0;0), '\n').unwrap_err();
        insta::assert_debug_snapshot!(state);
    }

    #[test]
    fn escape() {
        let state = StrState::default();
        let state = state.try_append_char(pos!(0;0), '\\').unwrap();
        insta::assert_debug_snapshot!(state);
    }

    #[test]
    #[should_panic]
    fn illigal_state() {
        let state = StrState::default();
        let state = state.try_append_char(pos!(0;0), '\\').unwrap();
        state.tokenize();
    }
}
