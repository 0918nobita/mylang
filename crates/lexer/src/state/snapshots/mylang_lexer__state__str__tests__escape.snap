---
source: crates/lexer/src/state/str.rs
assertion_line: 80
expression: state

---
StrState {
    start: 0:0,
    end: 0:0,
    escape: true,
    acc: "",
}
